src/core/mod.rs

```diff
@@ -21,34 +21,156 @@ use parsec_interface::operations::psa_key_attributes::KeyAttributes;
 use parsec_interface::operations::psa_sign_hash::Operation as PsaSignHash;
 use parsec_interface::operations::psa_verify_hash::Operation as PsaVerifyHash;
 use parsec_interface::operations::{NativeOperation, NativeResult};
-use parsec_interface::requests::Opcode;
-use parsec_interface::requests::ProviderID;
 use std::collections::HashSet;
 
+pub use parsec_interface::operations::{psa_algorithm, psa_key_attributes};
+pub use parsec_interface::requests::{BodyType, Opcode, ProviderID};
+
 /// Core client for Parsec service
 ///
-/// The client exposes low-level functionality for using the Parsec service
+/// The client exposes low-level functionality for using the Parsec service.
+/// Below you can see code examples for a few of the operations supported.
+///
+/// Creating a `CoreClient` instance:
+///```no_run
+///use parsec_client::auth::AuthenticationData;
+///use parsec_client::CoreClient;
+///
+///let app_name = String::from("app-name");
+///let app_auth_data = AuthenticationData::AppIdentity(app_name);
+///let client: CoreClient = CoreClient::new(app_auth_data);
+///```
+///
+/// Performing a Ping operation to determine if the service is available
+/// and what wire protocol it supports:
+///```no_run
+///# use parsec_client::auth::AuthenticationData;
+///# use parsec_client::CoreClient;
+///# let client: CoreClient = CoreClient::new(AuthenticationData::AppIdentity(String::from("app-name")));
+///let res = client.ping();
+///
+///if let Ok((wire_prot_v_maj, wire_prot_v_min)) = res {
+///    println!(
+///        "Success! Service wire protocol version is {}.{}",
+///        wire_prot_v_maj, wire_prot_v_min
+///    );
+///} else {
+///    panic!("Ping failed. Error: {:?}", res);
+///}
+///```
+///
+/// Providers are abstracted representations of the secure elements that
+/// PARSEC offers abstraction over. Providers are the ones to execute the
+/// cryptographic operations requested by the user.
+///
+/// Checking for available providers:
+///```no_run
+///# use parsec_client::auth::AuthenticationData;
+///# use parsec_client::CoreClient;
+///# let client: CoreClient = CoreClient::new(AuthenticationData::AppIdentity(String::from("app-name")));
+///use parsec_client::core::ProviderID;
+///
+///let desired_provider = ProviderID::Tpm;
+///let available_providers = client.list_providers().expect("Failed to list providers");
+///if available_providers
+///    .iter()
+///    .filter(|provider| provider.id == desired_provider)
```

Comment:
I am wondering if it makes sense to filter using the `id` field as this one might change in the future to be dynamic. Maybe one use-case could be that a client gets a list of all providers and check if one of them match with the UUID of the one he wants?

Reply:
Sounds good, shall replace!

Reply:
Now that I think of it, it makes sense to have some sort of mapping from `ProviderID` to the UUID, since that's the way to link the two constant enums together.... otherwise I'd just be saying "here, trust me, this is the correct UUID". It's much more elegant to do something like `ProviderID::Tpm.uuid()`

Reply:
The Provider ID might not be known a priori though, I think the flow would be the following for clients:
1. they call `list_providers`
2. they check if there is the UUID of the ones they are interested in (from [this](https://parallaxsecond.github.io/parsec-book/parsec_service/providers.html))
3. they now know the Provider ID of these providers and can make crypto op with them

I agree with a mapping but maybe a global mapping of "provider name" to UUID? Like:
"MbedCrypto" -> "1c1139dc-ad7c-47dc-ad6b-db6fdb466552"
something like that

Reply:
Added a `Provider` enum, instead of exposing `ProviderID`, and implemented methods to get the ID and the UUID for each one

Reply:
Great, nice!